tests/run-all.go

```diff
@@ -392,3 +397,112 @@ func saveLogToFile(file string, lines []string) {
 		}
 	}
 }
+
+// getE2eRegex gets the regex to filter which tests to run.
+// If E2E_TEST_REGEX is set, it overrides and uses that regex.
+// If not, it uses a config to build a regex. If the config is nil, it uses the default regex which is to run all tests.
+func getE2eRegex() (string, error) {
+	// if there's a regex, use it
+	e2eRegex := os.Getenv("E2E_TEST_REGEX")
+	if e2eRegex != "" {
+		return e2eRegex, nil
+	}
+
+	return buildRegexFromConfig(helper.KEDATestConfig)
+}
+
+// buildRegexFromConfig builds a regex string from a TestConfig
+func buildRegexFromConfig(config helper.TestConfig) (string, error) {
+	var regexParts []string
+
+	// For each known category, we get all the available tests under the tests/category directory.
+	// We then filter the tests we actually run based on the config exclude and includes.
+	// Then we incrementally build a regex based on the filtered tests.
+	for _, category := range config.GetAllCategories() {
+		availableTests, err := getAvailableTests("tests/" + category)
+		if err != nil {
+			return "", fmt.Errorf("error getting %q tests: %w", category, err)
+		}
+
+		var supportedTests []string
+		if categoryConfig, exists := config.TestCategories[category]; exists {
+			supportedTests = filterTests(availableTests, categoryConfig)
+		} else {
+			// use all tests in the category if no config is specified
+			supportedTests = availableTests
+		}
+
+		if len(supportedTests) > 0 {
+			// go regex doesn't support negative lookaheads, so we need to explicily include the tests we want to run
+			regexParts = append(regexParts, fmt.Sprintf("%s/(%s)/.*", category, strings.Join(supportedTests, "|")))
+		}
+	}
+
+	if len(regexParts) == 0 {
+		return ".*_test.go", nil
+	}
+
+	return fmt.Sprintf("^tests/(%s)_test\\.go$", strings.Join(regexParts, "|")), nil
+}
+
+// getAvailableTests returns all available test suites/directories for a category, as a slice of strings
+func getAvailableTests(categoryPath string) ([]string, error) {
+	entries, err := os.ReadDir(categoryPath)
+	if err != nil {
+		return nil, err
+	}
+
+	var tests []string
+	for _, entry := range entries {
+		if entry.IsDir() {
```

Comment:
I think that this should be `!entry.IsDir`, shouldn't?
Additionally, we can check if the file ends with `_test` because there are some helpers inside folders and probably we will include them in the regex. It doesn't hurt tbh, but just to reduce the regex

Reply:
That's true, we try to have a single test file per folder as we created package per tests, but there are folders like https://github.com/kedacore/keda/tree/main/tests/scalers/nats_jetstream where there are subfolders and not all of them are tests

```
├── scalers
│   ├── nats_jetstream
│   │   ├── helper
│   │   │   ├── nats_helper.go <- this is not a tests case
│   │   ├── nats_jetstream_cluster
│   │   │   ├── nats_jetstream_cluster_test.go
│   │   ├── nats_jetstream_standalone
│   │   │   ├── nats_jetstream_standalone_test.go
```

My suggestion was in the direction of filter files like `nats_helper_go`

Reply:
Please take a look at the latest commit, hopefully it answers your concern. It uses `filepath.WalkDir` and now specifically ignores those `helper` or `helpers` subdirectories.

But either way, I think `go test` ignores non `_test.go` anyways so in the end only the regex is affected like you said.

Reply:
> About filtering by top level folder, IDK because I don't have any example in mind, but I guess that some vendors could be interested in offer some other vendors but maybe they don't offer the full catalog, so big vendors like azure or AWS must be included or excluded, but can't be filtered as far as I understand, right?
I mean, currently there isn't any option to say that I want to include AWS CloudWatch e2e tests but not other like AWS dynamo, is it correct?

Just saw this comment, but it should be resolved now, I hope? :-)